rmf_site_editor/src/site/scenario.rs

```diff
@@ -18,148 +18,381 @@
-                    .removed_instances
-                    .iter_mut()
-                    .find(|e| **e == entity)
-                    .map(|e| e.clone());
-                if let Some(existing_removed_instance) = existing_removed_instance {
-                    current_scenario
-                        .moved_instances
-                        .retain(|(e, _)| *e != existing_removed_instance);
-                    current_scenario
-                        .added_instances
-                        .retain(|(e, _)| *e != existing_removed_instance);
-                    return;
+    // Insert instance modifier entities when new model instances are spawned and placed
+    if added_instances.is_empty() {
+        if !added_scenarios.is_empty() {
+            change_current_scenario.send(ChangeCurrentScenario(current_scenario_entity));
+        }
+        return;
+    }
+    let instance_modifier_entities =
+        get_instance_modifier_entities(current_scenario_entity, &children, &instance_modifiers);
+
+    for (instance_entity, new_pose) in added_instances.iter() {
+        if let Some((mut instance_modifier, _)) = instance_modifier_entities
+            .get(&instance_entity)
+            .and_then(|modifier_entity| instance_modifiers.get_mut(*modifier_entity).ok())
+        {
+            // If an instance modifier entity already exists for this scenario, update it
+            let instance_modifier = instance_modifier.as_mut();
+            match instance_modifier {
+                InstanceModifier::Added(_) => {
+                    *instance_modifier = InstanceModifier::new_added(new_pose.clone())
+                }
+                InstanceModifier::Inherited(_) => {
+                    *instance_modifier = InstanceModifier::new_inherited(Some(new_pose.clone()))
                 }
+                _ => {}
+            }
+        } else {
+            // If instance modifier entity does not exist in this scenario, spawn one
+            commands
+                .spawn(InstanceModifier::new_added(new_pose.clone()))
+                .insert(Affiliation(Some(instance_entity)))
+                .set_parent(current_scenario_entity);
+        }
 
-                let existing_added_instance: Option<&mut (Entity, Pose)> = current_scenario
-                    .added_instances
-                    .iter_mut()
-                    .find(|(e, _)| *e == entity);
-                if let Some(existing_added_instance) = existing_added_instance {
-                    existing_added_instance.1 = pose.clone();
-                    return;
-                } else if pose.is_added() {
-                    current_scenario
-                        .added_instances
-                        .push((entity, pose.clone()));
-                    return;
+        // Insert this new instance modifier into children scenarios as Inherited
+        let mut subtree_dependents = HashSet::<Entity>::new();
+        let mut queue = vec![current_scenario_entity];
```

Comment:
We need to add modifiers to all scenarios, not just children of the current scenario.